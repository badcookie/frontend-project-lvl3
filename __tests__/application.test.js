import fs from 'fs';
import path from 'path';
import nock from 'nock';
import timer from 'timer-promise';
import { html } from 'js-beautify';
import userEvent from '@testing-library/user-event';

import run from '../src/application';

nock.disableNetConnect();

const htmlOptions = {
  preserve_newlines: true,
  unformatted: [],
};

const fixturesPath = path.join(__dirname, '__fixtures__');

const readFixture = (filename) => {
  const filepath = path.join(fixturesPath, filename);
  return fs.readFileSync(filepath).toString();
};

const getTree = () => html(document.body.innerHTML, htmlOptions);

let elements;

describe('ui', () => {
  beforeEach(() => {
    const initHtml = readFixture('index.html');
    document.documentElement.innerHTML = initHtml;
    run();

    elements = {
      form: document.querySelector('form'),
      input: document.querySelector('input'),
      button: document.querySelector('button'),
      feeds: document.querySelector('.rss-feeds'),
    };
  });

  test('passing invalid url', async () => {
    expect(getTree()).toMatchSnapshot();

    await userEvent.type(elements.input, 'invalid url', { allAtOnce: true });
    elements.input.setAttribute('value', 'invalid url');

    await timer.start(10);
    expect(getTree()).toMatchSnapshot();

    elements.form.submit();
    await timer.start(10);

    expect(getTree()).toMatchSnapshot();
  });

  test('passing empty url', async () => {
    elements.form.submit();
    await timer.start(10);

    expect(getTree()).toMatchSnapshot();
  });

  test('valid url', async () => {
    const proxyUrl = 'https://cors-anywhere.herokuapp.com';

    const firstValidUrl = 'https://www.example.com/rss1.xml';
    const firstResponseRss = readFixture('rss1.xml');

    const secondValidUrl = 'https://www.example.com/rss2.xml';
    const secondResponseRss = readFixture('rss2.xml');

    nock(proxyUrl)
      .defaultReplyHeaders({ 'access-control-allow-origin': '*' })
      .get(`/${firstValidUrl}`)
      .reply(200, firstResponseRss);

    nock(proxyUrl)
      .defaultReplyHeaders({ 'access-control-allow-origin': '*' })
      .get(`/${secondValidUrl}`)
      .reply(200, secondResponseRss);

    await userEvent.type(elements.input, firstValidUrl, { allAtOnce: true });
    elements.input.setAttribute('value', firstValidUrl);

    await timer.start(10);
    elements.form.submit();

    await timer.start(10);
    await timer.start(10);
    await timer.start(10);
    await timer.start(10);

    expect(getTree()).toMatchSnapshot();

    await timer.start(10);

    await userEvent.type(elements.input, '', { allAtOnce: true });
    elements.input.setAttribute('value', '');

    await timer.start(10);
    expect(getTree()).toMatchSnapshot();

    await userEvent.type(elements.input, secondValidUrl, { allAtOnce: true });
    elements.input.setAttribute('value', secondValidUrl);

    await timer.start(10);
    elements.form.submit();

    await timer.start(10);
    await timer.start(10);
    await timer.start(10);
    await timer.start(10);

    expect(getTree()).toMatchSnapshot();

    // const feedLink = elements.feeds.querySelector('a');
    // await userEvent.click(feedLink);

    // expect(getTree()).toMatchSnapshot();
  });
});
