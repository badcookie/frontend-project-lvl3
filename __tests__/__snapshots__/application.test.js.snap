// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui passing empty url 1`] = `
"<div class=\\"container mt-5\\">
    <div class=\\"jumbotron\\">
        <h1>RSS Reader</h1>
        <p class=\\"lead\\">Start reading RSS today! It is easy, it is nicely.</p>
        <p class=\\"text-muted\\">https://ru.hexlet.io/lessons.rss</p>
        <form class=\\"form-inline rss-form\\">
            <div class=\\"form-group\\">
                <input type=\\"text\\" name=\\"url\\" class=\\"form-control\\" placeholder=\\"RSS link\\" style=\\"border-color: red;\\">
            </div>
            <button type=\\"submit\\" class=\\"btn btn-primary ml-1\\">Add</button>
        </form>
        <div class=\\"feedback text-danger\\">URL is required</div>
    </div>
    <div class=\\"row\\">
        <div class=\\"col rss-feeds\\"></div>
        <div class=\\"col rss-posts\\"></div>
    </div>
    <footer class=\\"footer\\">
        <p>© Hexlet Production</p>
    </footer>
</div>"
`;

exports[`ui passing invalid url 1`] = `
"<div class=\\"container mt-5\\">
    <div class=\\"jumbotron\\">
        <h1>RSS Reader</h1>
        <p class=\\"lead\\">Start reading RSS today! It is easy, it is nicely.</p>
        <p class=\\"text-muted\\">https://ru.hexlet.io/lessons.rss</p>
        <form class=\\"form-inline rss-form\\">
            <div class=\\"form-group\\">
                <input type=\\"text\\" name=\\"url\\" class=\\"form-control\\" placeholder=\\"RSS link\\">
            </div>
            <button type=\\"submit\\" class=\\"btn btn-primary ml-1\\">Add</button>
        </form>
    </div>
    <div class=\\"row\\">
        <div class=\\"col rss-feeds\\"></div>
        <div class=\\"col rss-posts\\"></div>
    </div>
    <footer class=\\"footer\\">
        <p>© Hexlet Production</p>
    </footer>
</div>"
`;

exports[`ui passing invalid url 2`] = `
"<div class=\\"container mt-5\\">
    <div class=\\"jumbotron\\">
        <h1>RSS Reader</h1>
        <p class=\\"lead\\">Start reading RSS today! It is easy, it is nicely.</p>
        <p class=\\"text-muted\\">https://ru.hexlet.io/lessons.rss</p>
        <form class=\\"form-inline rss-form\\">
            <div class=\\"form-group\\">
                <input type=\\"text\\" name=\\"url\\" class=\\"form-control\\" placeholder=\\"RSS link\\" value=\\"invalid url\\">
            </div>
            <button type=\\"submit\\" class=\\"btn btn-primary ml-1\\">Add</button>
        </form>
    </div>
    <div class=\\"row\\">
        <div class=\\"col rss-feeds\\"></div>
        <div class=\\"col rss-posts\\"></div>
    </div>
    <footer class=\\"footer\\">
        <p>© Hexlet Production</p>
    </footer>
</div>"
`;

exports[`ui passing invalid url 3`] = `
"<div class=\\"container mt-5\\">
    <div class=\\"jumbotron\\">
        <h1>RSS Reader</h1>
        <p class=\\"lead\\">Start reading RSS today! It is easy, it is nicely.</p>
        <p class=\\"text-muted\\">https://ru.hexlet.io/lessons.rss</p>
        <form class=\\"form-inline rss-form\\">
            <div class=\\"form-group\\">
                <input type=\\"text\\" name=\\"url\\" class=\\"form-control\\" placeholder=\\"RSS link\\" value=\\"invalid url\\" style=\\"border-color: red;\\">
            </div>
            <button type=\\"submit\\" class=\\"btn btn-primary ml-1\\">Add</button>
        </form>
        <div class=\\"feedback text-danger\\">Invalid URL</div>
    </div>
    <div class=\\"row\\">
        <div class=\\"col rss-feeds\\"></div>
        <div class=\\"col rss-posts\\"></div>
    </div>
    <footer class=\\"footer\\">
        <p>© Hexlet Production</p>
    </footer>
</div>"
`;

exports[`ui valid url 1`] = `
"<div class=\\"container mt-5\\">
    <div class=\\"jumbotron\\">
        <h1>RSS Reader</h1>
        <p class=\\"lead\\">Start reading RSS today! It is easy, it is nicely.</p>
        <p class=\\"text-muted\\">https://ru.hexlet.io/lessons.rss</p>
        <form class=\\"form-inline rss-form\\">
            <div class=\\"form-group\\">
                <input type=\\"text\\" name=\\"url\\" class=\\"form-control\\" placeholder=\\"RSS link\\" value=\\"https://www.example.com/rss1.xml\\">
            </div>
            <button type=\\"submit\\" class=\\"btn btn-primary ml-1\\" disabled=\\"\\">Add</button>
        </form>
    </div>
    <div class=\\"row\\">
        <div class=\\"col rss-feeds\\"></div>
        <div class=\\"col rss-posts\\"></div>
    </div>
    <footer class=\\"footer\\">
        <p>© Hexlet Production</p>
    </footer>
</div>"
`;

exports[`ui valid url 2`] = `
"<div class=\\"container mt-5\\">
    <div class=\\"jumbotron\\">
        <h1>RSS Reader</h1>
        <p class=\\"lead\\">Start reading RSS today! It is easy, it is nicely.</p>
        <p class=\\"text-muted\\">https://ru.hexlet.io/lessons.rss</p>
        <form class=\\"form-inline rss-form\\">
            <div class=\\"form-group\\">
                <input type=\\"text\\" name=\\"url\\" class=\\"form-control\\" placeholder=\\"RSS link\\" value=\\"\\">
            </div>
            <button type=\\"submit\\" class=\\"btn btn-primary ml-1\\">Add</button>
        </form>
    </div>
    <div class=\\"row\\">
        <div class=\\"col rss-feeds\\">
            <ul>
                <li>
                    <div><b>First lorem ipsum feed</b>
                        <p style=\\"font-size: 14px;\\">This is a constantly updating lorem ipsum feed</p>
                    </div>
                </li>
            </ul>
        </div>
        <div class=\\"col rss-posts\\">
            <div><a href=\\"http://example.com/test/1587843840\\">First post of 1st RSS</a><br></div>
            <div><a href=\\"http://example.com/test/1587843780\\">Second post of 1st RSS</a><br></div>
            <div><a href=\\"http://example.com/test/1587843720\\">Third post of 1st RSS</a><br></div>
            <div><a href=\\"http://example.com/test/1587843660\\">Fourth post of 1st RSS</a><br></div>
            <div><a href=\\"http://example.com/test/1587843600\\">Fifth post of 1st RSS</a><br></div>
        </div>
    </div>
    <footer class=\\"footer\\">
        <p>© Hexlet Production</p>
    </footer>
</div>"
`;

exports[`ui valid url 3`] = `
"<div class=\\"container mt-5\\">
    <div class=\\"jumbotron\\">
        <h1>RSS Reader</h1>
        <p class=\\"lead\\">Start reading RSS today! It is easy, it is nicely.</p>
        <p class=\\"text-muted\\">https://ru.hexlet.io/lessons.rss</p>
        <form class=\\"form-inline rss-form\\">
            <div class=\\"form-group\\">
                <input type=\\"text\\" name=\\"url\\" class=\\"form-control\\" placeholder=\\"RSS link\\" value=\\"https://www.example.com/rss2.xml\\">
            </div>
            <button type=\\"submit\\" class=\\"btn btn-primary ml-1\\">Add</button>
        </form>
        <div class=\\"feedback text-success\\">RSS has been loaded</div>
    </div>
    <div class=\\"row\\">
        <div class=\\"col rss-feeds\\">
            <ul>
                <li>
                    <div><b>First lorem ipsum feed</b>
                        <p style=\\"font-size: 14px;\\">This is a constantly updating lorem ipsum feed</p>
                    </div>
                </li>
                <li>
                    <div><a href=\\"#\\">Second lorem ipsum feed</a>
                        <p style=\\"font-size: 14px;\\">This is yet another constantly updating lorem ipsum feed</p>
                    </div>
                </li>
            </ul>
        </div>
        <div class=\\"col rss-posts\\">
            <div><a href=\\"http://example.com/test/1587843840\\">First post of 1st RSS</a><br></div>
            <div><a href=\\"http://example.com/test/1587843780\\">Second post of 1st RSS</a><br></div>
            <div><a href=\\"http://example.com/test/1587843720\\">Third post of 1st RSS</a><br></div>
            <div><a href=\\"http://example.com/test/1587843660\\">Fourth post of 1st RSS</a><br></div>
            <div><a href=\\"http://example.com/test/1587843600\\">Fifth post of 1st RSS</a><br></div>
        </div>
    </div>
    <footer class=\\"footer\\">
        <p>© Hexlet Production</p>
    </footer>
</div>"
`;
